diff --git a/posix.mak b/posix.mak
index 3235365..8efb4ab 100644
--- a/posix.mak
+++ b/posix.mak
@@ -35,6 +35,9 @@ endif
 DFLAGS = -I$(DRUNTIME_PATH)/import -I$(PHOBOS_PATH) \
 		 -L-L$(PHOBOS_PATH)/generated/$(OS)/release/$(MODEL) $(MODEL_FLAG)
 DFLAGS += -w -de
+ifeq (dragonflybsd,$(OS))
+        DFLAGS += -L-L/usr/local/lib
+endif
 
 # Default DUB flags (DUB uses a different architecture format)
 DUBFLAGS = --arch=$(subst 32,x86,$(subst 64,x86_64,$(MODEL)))
